### DISCUSSION TITLE
Ability to close tab when at beginning of history?

### DISCUSSION THREAD

**Discussion by PSeQXh** [qutebrowser/qutebrowser#7180] [Q&A]
Thanks for everyones work on qutebrowser!

Sometimes I find myself using just the mouse while surfing through a few pages. On these occasions I wish there was a way to allow qutebrowser to close a tab/ history chain when hitting the "At beginning of history" mark the second time (in attribution to opening a new tab/ history chain with middle mouse button press on any link).

The message seems to come from 
[qutebrowser/browser/eventfilter.py#L252](https://github.com/qutebrowser/qutebrowser/blob/46694ffd8b21ca2e2c1ecf99413abac0f75e01b2/qutebrowser/browser/eventfilter.py#L252).

Is there a way to allow qutebrowser to close the current tab on the next back mouse button press? (Maybe by storing the first hit in some kind of state variable, checking against a config flag, and if it's set calling some self._tab.close() function vs message.error()?)

Or is there already a way to close a tab by pressing a mouse button without the need to move the cursor to the tab bar, target the little rectangle of the current active tab while simultaniously pressing the middle mouse button?

**Comment by mschilli87** [comment]
I think this would conflict with my common usage pattern of browsing within a single tab and eventually quickly return to the initial page by typing `100H` or sth like that. With you suggestion, that could results in closing not only this tab, but also others as well, if I understoof correctly.

**Reply by PSeQXh** [reply]
Always interesting to see other common usage patterns :+1: 
When I'm using both hands I sometimes use `<Ctrl-H>` to accomplish the same.

In my mind the behavior would be bound specifically to hitting the "At beginning of history" message once and then deliberately initiating another action like pressing back mouse button again or typing another `H`. 
A quick test on wikipedia with approximately 10 items in history and entering `100H` resulted in message.error() shown only once. 
If that isn't the case in reality, and message.error() actually gets called 100-10=90 times but just is omitted, then maybe it helps to restrict this behavior to mouse button input only? 

So far I tried 
- setting a class variable in `eventfilter.py` and calling self._tab.close() (didn't work as method close() on _tab object doesn't exist, also `eventfilter.py` is probably not the right place to have access to tab's parent window for switching to previous or next tab)
- binding mouse buttons to commands with `:bind` (did not result in any change in behavior, probably because mouse button events get filtered? Also would not accomplish the only-if-beginning-of-history aspect)

I would imagine a config flag like `tabs.close_if_beginning_of_history` for it, so everyone of us is able to alter the behavior depending on our own common usage pattern :slightly_smiling_face:

**Reply by The-Compiler** [reply]
I think you should be able to do `self._tab.window_close_requested.emit()` - that uses the same logic as calling `window.close()` from JS does, i.e. (despite the name), closes the tab.

I imagine getting this to work properly will be quite some more effort. I'm assuming that e.g. opening other pages and then triggering an error again should *not* close the tab, but it would if just setting a flag...

I'm also not too happy to have this added to qutebrowser if I'm being honest, even as an option. It sounds too specific to your workflow, and also hard to get correct - I'd much rather make it possible to do something like this from the extension API (#30), but unfortunately, work on that has somewhat stalled as I'm working on more urgent stuff (mainly #5395).

As for a more generic way to make it possible to close tabs via the mouse, see #591.
